design-documents/services-decomposition-guidelines.md

```diff
@@ -0,0 +1,42 @@
+# Services decomposition guidelines
+
+## About
+
+Document contains general recommendations and principles that need be considered and followed during services decomposition.
+
+## Guidelines
+
+Services should have it's own core_config_data table that would contain settings specific to the service. Services should expose API to modify settings in core_config_data. Admin panel should support editing settings in the local database (settings for BFF) and for individual services via remote calls.
+
+Services should not have knowledge about stores/websites. As some services store information about stores/websites with it's data stores/websites may need to be validated. If validation is critical services can request information about stores/website from websites service to use for validation. If not, validation can be omitted.
+
+Modules use directory data (countries, cities, states). This data need to be reused between different services. As it almost never changes, it doesn't make much sense to add directory service. Directory data can be moved to configuration, so it can be shared across different services.
+
+Modules have dependency on session. Handling session should be responsibility of BFF. BFF can pass user data to the service if needed.
+
+For reporting purposes, services should publish event with data for reporting services (BFF after first iteration). Reporting service then can consume the data and aggregate it for reporting.
+
+Need to consider code reuse disadvantage for deployment when applying decomposition. Sharing code between different services means that when updating one service we would need to redeploy another service.
+
+Techniques to reduce number of calls
+* Calls can be grouped. Related data can be requested in one call and used in different places by the service.
+* Part of the logic can be moved in a separate scenario. For instance, on the shopping cart page we need to validate QTY for each item to make sure we have enough items in stock and product is in correct QTY. This operation can be executed on it’s own instead on each quote load.
+* Some of the operations became not very useful in multi service environment. For instance, websites/stores validation may not be very useful in checkout service and these calls can be eliminated.
+* Making service own the data. Even though we expose editing configuration on the BFF, each service should own it’s own configuration.
+* In some cases number of calls can be reduced by creating new service. For instance, creating service to calculate totals for cart allows to reduce number of calls compare to having totals calculation logic in checkout service and having individual collectors send requests to total collectors.
+
+Data interfaces can be reused across different services. For instance, ProductInterface can be used in Catalog and Checkout service. But it also has several disadvantages:
```

Comment:
May make sense to describe how these interfaces will be deployed and how implementations will look like on the server and the client.